ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -377,6 +426,79 @@ extendVRNew cfg tx
                       , vrLastTicketNo = nextTicketNo
                       }
 
+-- | Attempt to validate and add as many new transactions to the mempool as
+-- possible, returning the last 'ValidationResult' and the remaining
+-- transactions which couldn't be added due to the mempool capacity being
+-- reached.
+validateNew :: forall m blk. (MonadSTM m, ApplyTx blk)
+            => MempoolEnv m blk
+            -> [GenTx blk]
+            -> ValidationResult blk
+            -> STM m (ValidationResult blk, [GenTx blk])
+validateNew mpEnv txs res =
+  let res' = res { vrInvalid = [] }
+  in case txs of
+    []                  -> return (res', [])
+    headTx:remainingTxs -> do
+      -- First, attempt to individually validate the first new transaction.
+      --
+      -- If this fails due to the mempool capacity being met, we should simply
+      -- 'retry' as it will be cheap due to the fact that we've done very
+      -- little work in this STM transaction.
+      --
+      -- If, however, this is successful, we should continue to validate all
+      -- of the other new transactions one-by-one. If the mempool capacity
+      -- would be reached at any step, we update the 'InternalState' with the
+      -- work that we've already done and return the last 'ValidationResult'
+      -- along with the remaining unvalidated transactions.
+      --
+      -- @intricate: I'm actually not sure if it makes sense to do this
```

Comment:
Actually, it might make sense. A `retry` will be more efficient than constantly recursing since it would block until one of the `TVar`s that we've read from in this transaction has been updated.